app/org/sagebionetworks/bridge/services/ConsentServiceImpl.java

```diff
@@ -42,189 +46,191 @@
         this.sendMailService = sendMailService;
     }
     @Autowired
-    public final void setStudyConsentService(StudyConsentService studyConsentService) {
+    final void setStudyConsentService(StudyConsentService studyConsentService) {
         this.studyConsentService = studyConsentService;
     }
     @Autowired
-    public final void setUserConsentDao(UserConsentDao userConsentDao) {
+    final void setUserConsentDao(UserConsentDao userConsentDao) {
         this.userConsentDao = userConsentDao;
     }
     @Autowired
-    public final void setActivityEventService(ActivityEventService activityEventService) {
+    final void setActivityEventService(ActivityEventService activityEventService) {
         this.activityEventService = activityEventService;
     }
+    @Autowired
+    final void setStudyEnrollmentService(StudyEnrollmentService studyEnrollmentService) {
+        this.studyEnrollmentService = studyEnrollmentService;
+    }
+    @Autowired
+    final void setSubpopulationService(SubpopulationService subpopService) {
+        this.subpopService = subpopService;
+    };
     
     @Override
-    public ConsentSignature getConsentSignature(final Study study, final User user) {
-        checkNotNull(user, Validate.CANNOT_BE_NULL, "user");
-        checkNotNull(study, Validate.CANNOT_BE_NULL, "study");
+    public ConsentSignature getConsentSignature(Study study, SubpopulationGuid subpopGuid, User user) {
+        checkNotNull(study);
+        checkNotNull(subpopGuid);
+        checkNotNull(user);
         
         Account account = accountDao.getAccount(study, user.getEmail());
-        ConsentSignature signature = account.getActiveConsentSignature();
-        if (signature != null) {
-            return signature;
+        ConsentSignature signature = account.getActiveConsentSignature(subpopGuid);
+        if (signature == null) {
+            throw new EntityNotFoundException(ConsentSignature.class);    
         }
-        throw new EntityNotFoundException(ConsentSignature.class);
+        return signature;
     }
-
+    
     @Override
-    public User consentToResearch(final Study study, final User user, final ConsentSignature consentSignature,
-            final SharingScope sharingScope, final boolean sendEmail) {
-
+    public User consentToResearch(Study study, SubpopulationGuid subpopGuid, User user, ConsentSignature consentSignature, 
+            SharingScope sharingScope, boolean sendEmail) {
         checkNotNull(study, Validate.CANNOT_BE_NULL, "study");
+        checkNotNull(subpopGuid, Validate.CANNOT_BE_NULL, "subpopulationGuid");
         checkNotNull(user, Validate.CANNOT_BE_NULL, "user");
         checkNotNull(consentSignature, Validate.CANNOT_BE_NULL, "consentSignature");
         checkNotNull(sharingScope, Validate.CANNOT_BE_NULL, "sharingScope");
 
-        if (user.doesConsent()) {
+        ConsentStatus status = ConsentStatus.forSubpopulation(user.getConsentStatuses().values(), subpopGuid);
```

Comment:
Why iterate through the value set when you have a perfectly good map keyed off the subpopGuid that you already know?